### ISSUE TITLE
type_traits.h uses std::result_of

### ISSUE THREAD

**Issue by 7ofNine** [abseil/abseil-cpp#649]
abseil is not compilable with C++20. std::result_of has been deprecated since C++17 and is/will be removed for compliant std implementations. This creates compilation failures with the latest MSVC 16.6.0 preview 2. See std draft [diff.cpp17.depr] 7

**Comment by zhangxy988** [comment]
Thanks for reporting this. It seems that for C++17 onward we need to use `invoke_result_t`.

**Comment by yesudeep** [comment]
I think the removal made its way into the Visual C++ standard library implementation already and starts breaking now when using `/std:c++latest` with Abseil:

```
C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.26.28801/bin/HostX64/x64/cl.exe /nologo /DCOMPILER_MSVC /DNOMINMAX /D_WIN32_WINNT=0x0601 /D_CRT_SECURE_NO_DEPRECATE /D_CRT_SECURE_NO_WARNINGS /bigobj /Zm500 /EHsc /wd4351 /wd4291 /wd4250 /wd4996 /Iexternal/com_google_absl /Ibazel-out/x64_windows-fastbuild/bin/external/com_google_absl /showIncludes /MD /Od /Z7 /wd4117 -D__DATE__="redacted" -D__TIMESTAMP__="redacted" -D__TIME__="redacted" -DHAVE_BAZEL_BUILD /std:c++latest /W3 /DNOMINMAX /DWIN32_LEAN_AND_MEAN /D_CRT_SECURE_NO_WARNINGS /D_SCL_SECURE_NO_WARNINGS /D_ENABLE_EXTENDED_ALIGNED_STORAGE /bigobj /wd4005 /wd4068 /wd4180 /wd4244 /wd4267 /wd4503 /wd4800 /Fobazel-out/x64_windows-fastbuild/bin/external/com_google_absl/absl/container/_objs/hashtablez_sampler/hashtablez_sampler.obj /c external/com_google_absl/absl/container/internal/hashtablez_sampler.cc
Execution platform: @local_config_platform//:host
C:\tmp\yhlvxsm5\execroot\__main__\external\com_google_absl\absl/meta/type_traits.h(614): error C2039: 'result_of': is not a member of 'std'
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.26.28801\include\ctime(21): note: see declaration of 'std'
C:\tmp\yhlvxsm5\execroot\__main__\external\com_google_absl\absl/meta/type_traits.h(614): error C2653: 'result_of': is not a class or namespace name
C:\tmp\yhlvxsm5\execroot\__main__\external\com_google_absl\absl/meta/type_traits.h(614): error C2061: syntax error: identifier 'type'
C:\tmp\yhlvxsm5\execroot\__main__\external\com_google_absl\absl/meta/type_traits.h(616): error C2143: syntax error: missing ';' before '{'
C:\tmp\yhlvxsm5\execroot\__main__\external\com_google_absl\absl/meta/type_traits.h(616): error C2447: '{': missing function header (old-style formal list?)
```

**Comment by pepsiman** [comment]
Define _HAS_DEPRECATED_RESULT_OF=1 to allow your Visual C++ build to continue.

**Comment by chaubunhtha** [comment]
Yes

**Comment by jjpepper** [comment]
Is there a plan to actually fix this?

**Comment by derekmauro** [comment]
There is a pending internal change (cl/337927702) to fix this issue.

**Comment by jeremyong** [comment]
For those that ran into this like I did, the current issue with the fix in e19260f is that on MSVC, the `__cplusplus` version guard reports a very old standard by default, even when compiling with `/std:c++latest`. The solution is to ensure modules that link Abseil leverage the following compiler flag: https://docs.microsoft.com/en-us/cpp/build/reference/zc-cplusplus?view=msvc-160
